### PR TITLE
Fix build failure when kokkos +rocm is enabled.

### PR DESCRIPTION
**Pull request by srekolam**
when the + rocm is provided for the kokkos, we observe the below failure. 
CMake Error at cmake/Modules/FindTPLROCTHRUST.cmake:11 (FIND_PACKAGE):
     25      By not providing "Findrocthrust.cmake" in CMAKE_MODULE_PATH this project
     26      has asked CMake to find a package configuration file provided by
     27      "rocthrust", but CMake did not find one.

Looking at the cmake/ kokkos_tpls.:42 file
IF(KOKKOS_ENABLE_HIP AND NOT KOKKOS_HAS_TRILINOS)
  SET(ROCTHRUST_DEFAULT ON)

i added depends_on("rocthurst", + rocm)

### CODE REVIEW

**Review by eugeneswalker — DISMISSED** [review]
Thanks!

Does anyone else want to comment on this before we merge?

**Comment by cedricchevalier19** [comment]
> Looking at the cmake/ kokkos_tpls.:42 file IF(KOKKOS_ENABLE_HIP AND NOT KOKKOS_HAS_TRILINOS) SET(ROCTHRUST_DEFAULT ON)

This only means that the default behavior is to enable `rocthrust`; that is, the `Kokkos_ ENABLE_ROCTHRUST` option is enabled by default (https://kokkos.org/kokkos-core-wiki/keywords.html#third-party-libraries-tpls).
Another solution is to turn off this flag to CMake.

And even better, to add a variant that correctly passes the `Kokkos_ENABLE_ROCTHRUST` flag.

**Review by cedricchevalier19 — CHANGES_REQUESTED** [review]
Thank you for spotting this bug, and sorry for the inconvenience.

Adding a `rocthrust` variant for Kokkos >= 4.3 is a better solution.

Moreover, `spack` should explicitly pass the correct `Kokkos_ENABLE_ROCTHRUST` flag to CMake.

**Comment by cedricchevalier19** [comment]
This bug is also present for `onedpl` and `sycl`. I will submit the fix as a separate PR.

@nmm0 @lucbv @dalg24.

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 1]
```diff
@@ -212,6 +213,9 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("kokkos-nvcc-wrapper@master", when="@master+wrapper")
     conflicts("+wrapper", when="~cuda")
 
+    conflicts("+rocthrust", when="@4.3: ~rocm")
+    conflicts("~rocthrust", when="@4.3: +rocm")
```
Comment (dalg24) [inline var/spack/repos/builtin/packages/kokkos/package.py]: I am not so familiar with Spack but isn't registering the variant and doing this just a convoluted way of doing `depends_on("rocthrust", when="@4.3+rocm")` in the end?

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 2]
```diff
@@ -350,6 +354,7 @@ def cmake_args(self):
 
         if "+rocm" in self.spec:
             options.append(self.define("CMAKE_CXX_COMPILER", self.spec["hip"].hipcc))
+            options.append(self.define("Kokkos_ENABLE_ROCTHRUST", True))
```
Comment (dalg24) [inline var/spack/repos/builtin/packages/kokkos/package.py:358]: It is enabled by default unless the user explicitly disable it so I am not sure it is worth bothering with it.

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 1 above)
Comment (srekolam) [reply var/spack/repos/builtin/packages/kokkos/package.py]: i put that to address both the below- 
spack install kokkos + rocm +rocthrust
spack install kokkos +rocm

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 2 above)
Comment (srekolam) [reply var/spack/repos/builtin/packages/kokkos/package.py:358]: i meant to address the review comment above.

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 3]
```diff
@@ -213,6 +213,9 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("kokkos-nvcc-wrapper@master", when="@master+wrapper")
     conflicts("+wrapper", when="~cuda")
 
+    conflicts("+rocthrust", when="@4.3: ~rocm")
+    conflicts("~rocthrust", when="@4.3: +rocm")
```
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py]:
```suggestion
    conflicts("+rocthrust", when="@4.3: ~rocm")
```

~We do not have problem compiling kokkos without `rocthrust`.~

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 4]
```diff
@@ -351,6 +354,7 @@ def cmake_args(self):
 
         if "+rocm" in self.spec:
             options.append(self.define("CMAKE_CXX_COMPILER", self.spec["hip"].hipcc))
+            options.append(self.define("Kokkos_ENABLE_ROCTHRUST", True))
```
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py:357]:
```suggestion
            options.append(self.define_from_variant("Kokkos_ENABLE_ROCTHRUST", "rocthrust"))
```

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 5]
```diff
@@ -74,6 +74,7 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
         "hwloc": [False, "Whether to enable the HWLOC library"],
         "numactl": [False, "Whether to enable the LIBNUMA library"],
         "memkind": [False, "Whether to enable the MEMKIND library"],
+        "rocthrust": [False, "Whether to enable the thrust library for the HIP backend"],
```
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py]: ~It is a bit more complicated: `rocthrust` should be enabled by default when `+rocm` and disable else.~

**Review by cedricchevalier19 — COMMENTED** [review]
Comments on the use of the new `rocthrust` variant. 
However, I have asked the Kokkos team if we need a new variant or if we consider that spack users should always use `rocthrust` when using AMD GPUs.

**New**: please discard the following remarks. It is better to enable `rocthrust` always.

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 5 above)
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py]:
```suggestion
```

After more thought, always enabling `rocthrust` when enabling `room` makes more sense for the users. And it is more straightforward.

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 1 above)
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py]:
```suggestion
```

After more thought, always enabling `rocthrust` when enabling `room` makes more sense for the users. And it is more straightforward.

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 2 above)
Comment (cedricchevalier19) [reply var/spack/repos/builtin/packages/kokkos/package.py:358]: It is better to be explicit. And Kokkos can fail if `rocthrust` is not correctly found.

**Review by cedricchevalier19 — COMMENTED** [review]
After more thought, always enabling `rocthrust` when enabling `room` makes more sense for the users. And it is more straightforward.

Sorry for the noise and thank you for your work.

**Comment by eugeneswalker** [comment]
Is this ready now?

**Comment by cedricchevalier19** [comment]
@srekolam Can you please not introduce the `rocthrust` variant? Just add the dependency as it was done in your first commit. (and again, sorry for my first comments).

I tend to prefer that you still pass the Kokkos option to explicitly enable rocthrust.

**Comment by srekolam** [comment]
> @srekolam Can you please not introduce the `rocthrust` variant? Just add the dependency as it was done in your first commit. (and again, sorry for my first comments).
> 
> I tend to prefer that you still pass the Kokkos option to explicitly enable rocthrust.

i have reverted the change. can you please review and let me know if this is okay.

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 2 above)
Comment (dalg24) [reply var/spack/repos/builtin/packages/kokkos/package.py:358]: Why would it better to be explicit?
IMO this unnecessarily adds code to that package recipe.

**Review by dalg24 — APPROVED** [review]
I would drop the `-DKokkos_ENABLE_ROCTHRUST=ON` as it is not needed.
Not blocking though.

**Comment by eugeneswalker** [comment]
This fixes https://github.com/spack/spack/issues/44832 for me! Thanks!

```
...
==> Installing kokkos-4.3.01-l5bsidfdbr7d7qeiqcqxlft6nwfbpti2 [99/99]
==> No binary for kokkos-4.3.01-l5bsidfdbr7d7qeiqcqxlft6nwfbpti2 found: installing from source
==> Fetching https://github.com/kokkos/kokkos/archive/4.3.01.tar.gz
==> No patches needed for kokkos
==> kokkos: Executing phase: 'cmake'
==> kokkos: Executing phase: 'build'
==> kokkos: Executing phase: 'install'
==> kokkos: Successfully installed kokkos-4.3.01-l5bsidfdbr7d7qeiqcqxlft6nwfbpti2
  Stage: 2.55s.  Cmake: 10.43s.  Build: 15.68s.  Install: 1.19s.  Post-install: 0.32s.  Total: 30.67s
```

**Review by eugeneswalker — DISMISSED** [review]

**Comment by eugeneswalker** [comment]
@cedricchevalier19 Is this OK now? Can we merge?

**Review by cedricchevalier19 — APPROVED** [review]

**Review by cedricchevalier19 — APPROVED** [review]
LGTM

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 6]
```diff
@@ -198,6 +198,7 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
         variant(dev, default=dflt, description=desc)
     conflicts("+cuda", when="+rocm", msg="CUDA and ROCm are not compatible in Kokkos.")
     depends_on("intel-oneapi-dpl", when="+sycl")
+    depends_on("rocthrust", when="+rocm")
```
Comment (rbberger) [inline var/spack/repos/builtin/packages/kokkos/package.py]: Shouldn't this be
```suggestion
    depends_on("rocthrust", when="@4.3: +rocm")
```
?

Comment (dalg24) [reply var/spack/repos/builtin/packages/kokkos/package.py]: I can't speak to the Spack syntax but you are correct this dependency is only from Kokkos 4.3 release

Comment (cedricchevalier19) [reply var/spack/repos/builtin/packages/kokkos/package.py]: You are true, the requirement is only from version 4.3.

**Comment by srekolam** [comment]
@eugeneswalker / @tldahlgren  , this PR is ready for review/merge. Can you please let me know if this can be approved.

**Review by eugeneswalker — APPROVED** [review]

**Pull request merged**